py/private/py_executable.bzl

```diff
@@ -0,0 +1,93 @@
+"""Some vendored helper code, copied from Bazel:
+https://github.com/bazelbuild/bazel/blob/37983b2f89cca7cc8b0bd596f4558fa36c8fbff2/src/main/starlark/builtins_bzl/common/python/py_executable.bzl
+
+There's not a way for us to reference that at runtime (via @bazel_tools, say) and even if there was
+we don't want to have to detect the version of Bazel that the user is running to know whether code
+we want to re-use will be present.
+"""
+
+def csv(values):
+    """Convert a list of strings to comma separated value string."""
+    return ", ".join(sorted(values))
+
+def _path_endswith(path, endswith):
+    # Use slash to anchor each path to prevent e.g.
+    # "ab/c.py".endswith("b/c.py") from incorrectly matching.
+    return ("/" + path).endswith("/" + endswith)
+
+def _determine_main(ctx):
+    """Determine the main entry point .py source file.
+
+    Args:
+        ctx: The rule ctx.
+
+    Returns:
+        Artifact; the main file. If one can't be found, an error is raised.
+    """
+    if ctx.attr.main:
+        # Deviation from rules_python: allow a leading colon, e.g. `main = ":my_target"`
+        proposed_main = ctx.attr.main.removeprefix(":")
+        if not proposed_main.endswith(".py"):
+            fail("main must end in '.py'")
```

Comment:
nit: it would be nice if the target was included in these error messages, makes it much easier to find in a large repo.